lib/nimble_totp.ex

```diff
@@ -185,14 +230,40 @@ defmodule NimbleTOTP do
   In this example by validating first against the current time, but also against 30 seconds
   ago, we allow the _previous_ code, to be still valid.
   """
+  @spec valid?(binary(), [option() | validate_option()]) :: boolean()
   def valid?(secret, otp, opts \\ [])
 
   def valid?(secret, <<a1, a2, a3, a4, a5, a6>>, opts) do
-    <<e1, e2, e3, e4, e5, e6>> = verification_code(secret, opts)
+    time = opts |> Keyword.get(:time, System.os_time(:second)) |> to_unix()
+    period = Keyword.get(opts, :period, @default_totp_period)
+
+    <<e1, e2, e3, e4, e5, e6>> = verification_code(secret, time, period)
 
     (bxor(e1, a1) ||| bxor(e2, a2) ||| bxor(e3, a3) ||| bxor(e4, a4) ||| bxor(e5, a5) |||
-       bxor(e6, a6)) === 0
+       bxor(e6, a6)) === 0 and not reused?(time, period, opts)
   end
 
   def valid?(_secret, _otp, _opts), do: false
+
+  @spec reused?(integer(), pos_integer(), [option() | validate_option()]) :: boolean()
+  defp reused?(time, period, opts) do
+    if since = opts |> Keyword.get(:since) |> to_unix() do
+      Integer.floor_div(time, period) <= Integer.floor_div(since, period)
+    else
+      false
+    end
+  end
+
+  @spec to_unix(nil) :: nil
+  defp to_unix(nil), do: nil
+
```

Comment:
```suggestion
    if since = Keyword.get(opts, :since) do
      Integer.floor_div(time, period) <= Integer.floor_div(to_unix(since), period)
    else
      false
    end
  end

```